Classes/BITChannel.m

```diff
@@ -123,39 +128,61 @@ - (BOOL)isQueueBusy {
   return self.channelBlocked;
 }
 
-- (void)persistDataItemQueue {
+- (void)persistDataItemQueue:(char **)string {
   [self invalidateTimer];
-  if (!BITSafeJsonEventsString || strlen(BITSafeJsonEventsString) == 0) {
+  char* jsonStream = NULL;
+  char *newEmptyString = NULL;
+  do {
+    jsonStream = *string;
+    newEmptyString = strdup("");
+    
+    // Reset both, the async-signal-safe buffer and item counter. Swaps jsonStream and BITSafeJsonEventsString.
+    if (OSAtomicCompareAndSwapPtr(jsonStream, newEmptyString, (void*)string)) {
+      self.dataItemCount = 0;
```

Comment:
This is not sufficient, because `self.dataItemCount` is being guarded by `@synchronized()` in other places - Add `@synchronized(self){ }` around the reset of `self.dataItemCount` to make the setter call atomic.

Reply:
I agree with Gwynne since this method is called from the `mainQueue` and the `dataItemsOperations` queue.